CIP7 - Message signing

SebastienGllmt opened the pull request:
Private keys can be used to sign arbitrary data. If you have the public key, you can verify the data was signed by the owner of the private key. This is how transaction signing works internally but its utility is not limited to transactions. This document tries to set a standard for how to represent and verify signed messages for Cardano.

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+
+Note that we do NOT follow the COSE specification exactly for the following reason:
+
+- We need to restrict the number of algorithms used to only ones that would already be required to provide base Cardano functionality (see E1).
```
I'd appreciate it if in the following text it was very clearly indicated which parts come from the COSE spec and which deviate. I'd be tempted to just omit most of the content that exactly follows COSE, perhaps.

KtorZ replied:
Using a subset of COSE would be fine I think, but it has to be _implementable_ from pure COSE. So that any library implementing COSE could be used to implement this spec.

michaelpj commented on MessageSigning/MessageSigning.md:
````diff
@@ -0,0 +1,437 @@
+
+```
+? address: bstr,
+? address_pubkey: bstr,
````
For a different application, I am interested in the case where A provides B with a signature of a piece of data, but B does not (yet) know A's public key, so cannot actually perform verification. So we want to "annotate" the signature with the key that provided it (under some circumstances this might be an information leak, but not for this application).

However, in this case the public key need not correspond to an address. So my suggestion would simply be to have this header be `signing_pubkey` instead, which hopefully doesn't affect the clarity in the case where it also corresponds to an address, but would make more sense for non-address-related usecases.

michaelpj replied:
Moreover, reading through some of the COSE spec, it *seems* to me that there are already appropriate places for putting this information. For example, the `kid` unprotected header:
```
   kid:  This parameter identifies one piece of data that can be used as
      input to find the needed cryptographic key.  The value of this
      parameter can be matched against the 'kid' member in a COSE_Key
      structure.  Other methods of key distribution can define an
      equivalent field to be matched.  Applications MUST NOT assume that
      'kid' values are unique.  There may be more than one key with the
      same 'kid' value, so all of the keys associated with this 'kid'
      may need to be checked.  The internal structure of 'kid' values is
      not defined and cannot be relied on by applications.  Key
      identifier values are hints about which key to use.  This is not a
      security critical field.  For this reason, it can be placed in the
      unprotected headers bucket.
```
In particular, it seems like it would be conformant to put the signing key itself in the `kid` header.

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+1) P1 - (optional) knowledge of the relation of a public key and a Cardano address
+1) P2 - Knowledge of which algorithm was used to sign
+
+For `P1`, just knowing the public key is not sufficient to know the address (which we need to know if we want to prove ownership of a UTXO or transaction). To resolve this, first we (optionally) add to the header an extra field.
```
Why can't the sender put the address in the signed message if they need to convey it?

michaelpj replied:
In fact, I'm a little confused by the rationale here. During transaction validation, all that is required to prove ownership of a UTXO is (a signature from) the corresponding public key. Why do we need more here?

KtorZ replied:
@SebastienGllmt when it comes purely to the signing / verification procedure, it is purely Ed25519 even for legacy addresses. The chaincode and attributes are needed in order to re-compute the address root so that one can check that the signature was produced by the right private key. 

I am not sure to understand the full rationale here then. If we're looking for at message signing from a more general perspective, then the public key itself is sufficient to verify a signature made from a "legacy" private key.

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+
+1) We want to limit the number of allowed signing algorithms to exactly the set used in Cardano because of `E2`.
+1) There is a 1-1 mapping between the pair `address`, `address_meta` and what signing algorithm is used and so this information would be redundant.
+1) Allowed values in the `alg` field must be present in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml#key-type) (and new entries have to be manually approved). This is not sufficient for us as we often have to use cutting edge cryptography specialized for our domain.
```
I don't find this reasoning convincing:
1. Why not say that we use the header, but restrict which values are allowed for the algorithm?
2. This is true, but it seems like a small amount of redundant information.
3. This is a genuine problem. However, we've had success in getting entries added to such standards in the past (e.g. we've had new CBOR tags added). Has anyone reached out to the standards body in question? Even if we have to use a non-standard entry here, it would IMO be preferable to follow the standard format in all other regards. At least then anyone who feeds this to a standard library will get a comprehensible error ("unknown encryption algorithm" or something) rather than a parse error.

It seems like quite a cost to *almost* follow the COSE spec and then deviate at the last minute. That means that no standard tool or library that works with COSE will be able to handle these messages in the end.

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+1) There is a 1-1 mapping between the pair `address`, `address_meta` and what signing algorithm is used and so this information would be redundant.
+1) Allowed values in the `alg` field must be present in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml#key-type) (and new entries have to be manually approved). This is not sufficient for us as we often have to use cutting edge cryptography specialized for our domain.
+
+### Encryption
```
I haven't reviewed this in detail because it's hard to tell which parts are following the COSE spec and which aren't. (As you've probably guessed, I would really like us to deviate from the spec as little as possible!)

SebastienGllmt replied:
The `ChachaPoly` encryption is the same as described in https://tools.ietf.org/html/rfc8152#section-10.3 but I used different parameters than they did (but there is no reason we can't use the same parameters as the COSE spec if we want to use the `alg` field

The `Ed25519PubKey` just described how to use a share public key to power the ChachaPoly option so it works the same way regardless of whether or not we decide to use the `alg` field. Although the COSE spec doesn't mention anything about using an elliptic curve to generate the password, we have a cryptography auditor agree that this approach is sound. However, It's unfortunate we have to define a custom header to get this working instead of this being part of the COSE spec

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+1) The size of the raw `payload` would otherwise be too big to fit in hardware wallet memory (see E1). Note that the exact size for which this is the case depends on the device.
+1) The payload characters (ex: non-ASCII) that cannot be displayed on the hardware wallet device (see E3)
+
+We RECOMMEND showing the user the full payload on the device if possible because it lowers the attack surface (otherwise the user has to trust that the hash of the payload was calculated correctly).
```
Why? Can't the wallet verify this for them?

michaelpj replied:
If nobody is verifying the hash then it seems like you lose a lot of security properties. What if an attacker switches the payload?

michaelpj commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,437 @@
+
+### User-Facing Encoding
+
+Once we have our top-level `encrypted_message` or `signed_message` we need to encode them in way that can be displayed to users (doesn't need to be stored and can be inferred just from the data)
```
Why does this need to be standardised? Can't wallets chose how to do this as appropriate?

michaelpj reviewed (commented):
I'd like to see this work! My main concern is that we follow the COSE spec as much as possible. This will really ease interoperability: nobody wants to reimplement a subtly different version of COSE!

KtorZ commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,433 @@
+Although `v3` is relevant to us for encoding public keys into addresses, we do not use `bech32`'s scheme for encoding in this specification. This is because
+
+- The payload may be too big to reasonably encode in a QR code so the benefits of using base32 are limited.
+- BCH checksums are not made for large payloads and additionally the polynomial used has to be fine-tuned for the expected length (but the length of our payload varies too much in this spec)
```
It's good to have context but perhaps a bit superfluous / irrelevant to this particular CIP? What about making this much shorter?

KtorZ commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,433 @@
+---
+CIP: TBD
```
I'd be happy to assign this one a number already so we can reference it in the upcoming CIP about multisig wallets.

refi93 commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,433 @@
+
+We RECOMMEND showing the user the full payload on the device if possible because it lowers the attack surface (otherwise the user has to trust that the hash of the payload was calculated correctly).
+
+`Blake2b224` was chosen specifically because `224` bits is already a long string for hardware wallets.
```
If we allow signing the hash of the transaction directly by the hardware wallet without making the device hash the payload by itself I see the risk of the user unintentionally signing a valid transaction hash without them being aware. The user may even see the hash on the device but how does he know for sure it isn't a valid transaction hash?

A possible workaround could be appending a header to the hash to be signed so it does not have the same structure as a valid tx hash, thus making the attack (at least as I described it) impossible.

refi93 replied:
indeed, digest length for tx hashes is 256, not 224

SebastienGllmt replied:
When I wrote this spec, Cardano had a feature (in both Byron and Jormungandr) where you would sign objects with a specific prefix to avoid any chance of collision like this (ex: transactions would be signed as ('x' || payload), etc.) but this was removed in Haskell Shelley.
This message signing spec used to reserve a new prefix for signed messages, but since Cardano removed this prefix feature, I removed the prefix from this spec also. It's entirely possible passing just the hash to the hardware wallet no longer works after this change (and you would instead have to pass the full body and calculate the hash on the device itself)

SebastienGllmt replied:
I reviewed this again and I think it's still fine as-is because you're signing the payload wrapped in a structure so the probability of a collision is basically 0.

dcoutts reviewed (commented):
@SebastienGllmt I asked IOHK's internal auditors to review this CIP. The following is their initial review.

- In the spec it seems the message is signed then encrypted. However, it should be the other way around (encrypt then sign), because we would want to authenticate the encrypted message. Perhaps I’m reading this incorrectly? 

- It should be an essential requirement to make sure that signing/verification keys are only used for one single signing algorithm. Similarly, encryption/decryption keys should only be used for one single encryption algorithm (Though I think this is satisfied in the spec, it should be an explicit requirement, as breaking it could be detrimental to security).

- In section “Existing Code”, it mentions a replay attack where a person could take the signature for “BOB” and give it to someone else. However, usually for challenge-response type of authentication the challenger asks for a random string to be signed. So I believe this should not be too much of a concern.

- Having the address/public key as optional: This seems a bit strange, because this means that the signature would be valid/invalid based on the (assumed) knowledge of the receiver. Eg. if the receiver knows the public/verification key then the signature is valid; otherwise it is not.

- How does this scheme work with eg. HD wallets? Are the keys going to be one-time use?

- In section “Encryption”:    Is the encryption key/password the same type as the encrypted msg/ciphertext? (This seems to be the case, as both are defined as “ciphertext: bstr/nil”) 

- In Step 4 of the “Public-key based encryption” section: Why is it necessary to hash the secret, instead of just sharing the hashed value?

KtorZ commented on MessageSigning/MessageSigning.md:
```diff
@@ -0,0 +1,433 @@
+
+Although COSE defines multiple ways to encrypt, we simplify our spec to the two following cases:
+
+1) Encrypted with the recipient's public key (called `key transport` in COSE spec)
```
How would that work in practice? I don't recall seeing any ed25519 library which allowed public-key encryption. They only typically allow private-key signing. So I wonder if it's even possible to do such thing with Ed25519? 

COSE speaks about Diffie-Hellman and RSA is also a possibility. I found a few things in the literature using ECDSA which could suggest that this is achievable for Ed25519 curves but that's definitely something to ask cryptographers about?

SebastienGllmt replied:
How to handle ed25519 keys is explained in the header "Public key based encryption"

KtorZ replied:
Yup sorry, I read through that later on and saw how the document defines it with Diffie-Hellman shared secret, should've removed the comment :+1:

dcoutts reviewed (approved):
@SebastienGllmt has made changes to address the feedback from the IOHK internal auditors. He's said he'll add a comment to say point-by-point how the feedback was addresses (or why no changes were needed). On that basis I'm happy to move to draft status.

SebastienGllmt commented:
I addressed & responded  to all the comments by the cryptographer

> In the spec it seems the message is signed then encrypted. However, it should be the other way around (encrypt then sign), because we would want to authenticate the encrypted message. Perhaps I’m reading this incorrectly?

Although COSE allows both ways (`encrypt then sign` and `sign then encrypt`), `sign then encrypt` is more flexible and takes less bytes on the hardware wallet, so I would prefer to make it the default. We can expand the CIP to explicitly mention other combinations if we feel there is a need in the future

> It should be an essential requirement to make sure that signing/verification keys are only used for one single signing algorithm. Similarly, encryption/decryption keys should only be used for one single encryption algorithm (Though I think this is satisfied in the spec, it should be an explicit requirement, as breaking it could be detrimental to security).

Although I agree, currently it's not possible by construction for this to happen in this spec so I'm not sure if it's worth mentioning. I made which encryption algorithm to use more explicit since this comment was made to account for this. I also think COSE does mention this somewhere though.

> In section “Existing Code”, it mentions a replay attack where a person could take the signature for “BOB” and give it to someone else. However, usually for challenge-response type of authentication the challenger asks for a random string to be signed. So I believe this should not be too much of a concern.

challenge-response is not easy to do from a smart contract because there is no source of randomness that can be introdued within blocks (only between blocks through an orcale or something similar) which is why I explicitly mention dapps in the replay attack section

> Having the address/public key as optional: This seems a bit strange, because this means that the signature would be valid/invalid based on the (assumed) knowledge of the receiver. Eg. if the receiver knows the public/verification key then the signature is valid; otherwise it is not.

I've re-worked this section since this comment was made so I think it should be addressed.

> How does this scheme work with eg. HD wallets? Are the keys going to be one-time use?

You sign with the payment keys for the UTXO entries you want to prove ownership of.
It's not one-time use in the sense that to prove ownership, you sign a message with the same key you would also sign to spend the UTXO entry.
However, this is not something specific to HD wallets (you could easily have a raw private key that doesn't come from a recovery phrase that you want to prove ownership of)

> In section “Encryption”:    Is the encryption key/password the same type as the encrypted msg/ciphertext? (This seems to be the case, as both are defined as “ciphertext: bstr/nil”)

Yes, that's how it's defined in COSE.

> In Step 4 of the “Public-key based encryption” section: Why is it necessary to hash the secret, instead of just sharing the hashed value?

Yes, you're right that it'sn ot necessary. I've removed this.

ilap commented:
My comment in the Cardano forum on the encryption part this CIP.
https://forum.cardano.org/t/message-signing-specification/41032/3
